### PR TITLE
Fix number regex to solve #1909

### PR DESCRIPTION
**Pull request by kbenoit**
We have a long, clever regex to get the cases in the added test, but it didn't cover plain numbers of 10000 or over!!

Solves #1909

### CODE REVIEW

File: R/tokens.R [hunk 1]
```diff
@@ -483,7 +483,7 @@ compile_removals_regex <- function(remove_separators = FALSE,
     }
     if (remove_numbers) {
         # includes currency amounts and those containing , or . digit separators
-        regex_to_remove <- c(regex_to_remove, "^\\p{Sc}{0,1}([0-9]{1,3}([,.][0-9]{3})*([,.][0-9]+)?|[.,][0-9]+)\\p{Sc}{0,1}$")
+        regex_to_remove <- c(regex_to_remove, "^\\p{Sc}{0,1}([0-9]{1,3}([,.][0-9]{3})*([,.][0-9]+)?|[.,][0-9]+)\\p{Sc}{0,1}$|^\\d+$")
```
Comment (koheiw) [inline R/tokens.R]: How about sicking to Unicode character class `\\p{N}`? Also how about simplifying it to `"^\\p{Sc}?[,.\\p{N}]+$"`?

Comment (kbenoit) [reply R/tokens.R]: Simplifying is good, I agree, so is using the Unicode N character class. But we need also to check that:
- the `,` or `.` are thousands or decimal places, properly formed (groups of three unless a decimal)
- currency symbol can come at the end as well

**Comment by koheiw** [comment]
It seems that there currency signs are separated in our tokenizer. I think this is good.
``` r
txt <- "
$100,000
₣100,000
$100,000
¥100,000
$100,000
₡100,000
£100,000
"
quanteda::tokens(txt)
#> Tokens consisting of 1 document.
#> text1 :
#>  [1] "$"       "100,000" "₣"       "100,000" "$"       "100,000" "¥"      
#>  [8] "100,000" "$"       "100,000" "₡"       "100,000"
#> [ ... and 2 more ]
```

**Comment by koheiw** [comment]
Here is a list of top 1000 tokens that contain number in newspapers. We might need to decide which types of tokens `remove_number` should remove.

Definitely pure numbers (1-57). "*m" and "*bn", "am", "pm" would be nice to remove, but not ""19-year-old" or "co2".
```r
> names(topfeatures(dfm(toks, select = "[\\p{N}]+", valuetype = "regex"), 1000))
   [1] "2016"         "10"           "2015"         "2014"         "20"           "15"           "30"          
   [8] "1"            "2013"         "50"           "2"            "12"           "100"          "2012"        
  [15] "11"           "18"           "4"            "5"            "25"           "2008"         "40"          
  [22] "7"            "22"           "3"            "23"           "24"           "2010"         "17"          
  [29] "2011"         "13"           "14"           "9"            "19"           "6"            "16"          
  [36] "21"           "8"            "29"           "28"           "2009"         "60"           "27"          
  [43] "2017"         "26"           "2020"         "1,000"        "80"           "2007"         "31"          
  [50] "90"           "35"           "200"          "70"           "500"          "45"           "2005"        
  [57] "2006"         "100,000"      "10,000"       "300"          "48"           "2003"         "2000"        
  [64] "2004"         "0.5"          "36"           "1990s"        "1980s"        "1m"           "1bn"         
  [71] "2,000"        "55"           "3,000"        "2018"         "32"           "5,000"        "38"          
  [78] "75"           "2001"         "42"           "2002"         "33"           "150"          "34"          
  [85] "1,237"        "37"           "20,000"       "43"           "400"          "1970s"        "53"          
  [92] "2030"         "49"           "65"           "44"           "39"           "1992"         "1999"        
  [99] "52"           "1998"         "600"          "47"           "50,000"       "1.5"          "41"          
 [106] "1997"         "30,000"       "1996"         "4,000"        "1994"         "51"           "0.3"         
 [113] "0.4"          "100m"         "2.5"          "46"           "1995"         "40,000"       "250"         
 [120] "54"           "56"           "500,000"      "200,000"      "0.6"          "0.2"          "130"         
 [127] "800"          "2bn"          "95"           "10m"          "58"           "6,000"        "1960s"       
 [134] "72"           "21st"         "57"           "62"           "67"           "66"           "64"          
 [141] "1990"         "1,500"        "g4s"          "1993"         "2m"           "59"           "g20"         
 [148] "85"           "300,000"      "3bn"          "10bn"         "15,000"       "350"          "25,000"      
 [155] "1989"         "0.1"          "500m"         "2019"         "73"           "63"           "8am"         
 [162] "69"           "250,000"      "1.3"          "10-year"      "80,000"       "61"           "1.2"         
 [169] "11,000"       "700"          "1991"         "0.8"          "0.7"          "7pm"          "1975"        
 [176] "99"           "20m"          "3m"           "2025"         "68"           "4bn"          "74"          
 [183] "20th"         "1988"         "1.4"          "1.5bn"        "200m"         "5m"           "2014-15"     
 [190] "10am"         "1980"         "1.7"          "co2"          "3.5"          "1984"         "7,000"       
 [197] "15-year-old"  "400,000"      "mi6"          "2,500"        "1979"         "90s"          "71"          
 [204] "1983"         "80s"          "150,000"      "2.2"          "12,000"       "g7"           "1985"        
 [211] "120"          "1.1"          "50m"          "77"           "20s"          "5bn"          "1.6"         
 [218] "2050"         "1987"         "30m"          "350m"         "8,000"        "1950s"        "1976"        
 [225] "8pm"          "6pm"          "130m"         "9pm"          "76"           "16-year-old"  "4.5"         
 [232] "600,000"      "hs2"          "1972"         "1.2bn"        "60,000"       "10pm"         "2015-16"     
 [239] "1973"         "40m"          "98"           "11am"         "1986"         "1,200"        "35,000"      
 [246] "5pm"          "400m"         "1.3bn"        "70,000"       "17-year-old"  "2.3"          "78"          
 [253] "81"           "19th"         "9,000"        "2c"           "300m"         "1982"         "1.9"         
 [260] "9am"          "24-hour"      "1.5m"         "93"           "20bn"         "70s"          "1.8"         
 [267] "6bn"          "84"           "96"           "180"          "1974"         "3d"           "9.30am"      
 [274] "4m"           "19-year-old"  "mi5"          "7m"           "q1"           "86"           "88"          
 [281] "18-year-old"  "2021"         "1970"         "2.6"          "900"          "1930s"        "2.4"         
 [288] "140"          "83"           "8bn"          "1981"         "1968"         "800,000"      "45,000"      
 [295] "87"           "94"           "97"           "1971"         "15m"          "23-year-old"  "60s"         
 [302] "2pm"          "0.9"          "2013-14"      "30s"          "6m"           "1964"         "1967"        
 [309] "3pm"          "20-year-old"  "18th"         "2.1"          "4,300"        "1pm"          "1,400"       
 [316] "911"          "1945"         "82"           "267"          "999"          "89"           "2,383"       
 [323] "2016-17"      "6.5"          "2000s"        "100bn"        "11th"         "1978"         "0"           
 [330] "30bn"         "18,000"       "2.7"          "700,000"      "7am"          "1977"         "79"          
 [337] "15bn"         "ms804"        "11.30am"      "22-year-old"  "24-year-old"  "250m"         "92"          
 [344] "6am"          "91"           "1.4bn"        "90,000"       "110"          "27-year-old"  "4pm"         
 [351] "60m"          "q4"           "25m"          "3.5bn"        "25-year-old"  "13,000"       "26-year-old" 
 [358] "10.30am"      "18c"          "14,000"       "2019-20"      "10th"         "3,500"        "0.25"        
 [365] "12.30pm"      "2am"          "2.5bn"        "18bn"         "5.5"          "571m"         "3.30pm"      
 [372] "7bn"          "21-year-old"  "2nd"          "14-year-old"  "1,600"        "101"          "6.7"         
 [379] "2.8"          "29-year-old"  "1.3m"         "1969"         "70m"          "12-year-old"  "1922"        
 [386] "1,300"        "14m"          "700m"         "1961"         "1.1bn"        "74-year-old"  "75,000"      
 [393] "150m"         "1948"         "2017-18"      "28-year-old"  "1.2m"         "270"          "50s"         
 [400] "5am"          "120,000"      "1916"         "30-year-old"  "450"          "5.1"          "11m"         
 [407] "1.5c"         "bbc1"         "50bn"         "12bn"         "3.6"          "10-year-old"  "3.7"         
 [414] "34-year-old"  "22,000"       "2.30pm"       "1959"         "8m"           "7.5"          "600m"        
 [421] "2.5m"         "35m"          "1965"         "170"          "360"          "9m"           "50p"         
 [428] "12m"          "17,000"       "11.5m"        "2.3bn"        "111"          "16,000"       "130,000"     
 [435] "3am"          "a320"         "40s"          "35-year-old"  "4.9"          "7.58am"       "4.30pm"      
 [442] "123"          "2.9"          "mhl0"         "30-year"      "100th"        "1963"         "ps4"         
 [449] "3.54pm"       "3.3"          "3.4"          "m5s"          "45-year-old"  "1st"          "e3"          
 [456] "1,100"        "116"          "11pm"         "2020-21"      "800m"         "13-year-old"  "2022"        
 [463] "160,000"      "n1"           "40-year-old"  "9gu"          "160"          "g2"           "no2"         
 [470] "125"          "3.2"          "15th"         "1.8bn"        "10.15am"      "3.16pm"       "750"         
 [477] "38,000"       "1960"         "6.30pm"       "1956"         "4.5bn"        "175"          "8.29am"      
 [484] "230"          "31-year-old"  "3.9"          "1.9bn"        "12th"         "90th"         "hb2"         
 [491] "1,700"        "4.3"          "2.02pm"       "1.7bn"        "55,000"       "8.11am"       "21,000"      
 [498] "3.8"          "9bn"          "1940s"        "27,000"       "80m"          "40bn"         "33,000"      
 [505] "8.40am"       "1000"         "3.44pm"       "4am"          "9.38am"       "8.23am"       "4.4bn"       
 [512] "180,000"      "1951"         "2.43pm"       "1.4m"         "8.30am"       "11-year-old"  "2.15pm"      
 [519] "3.50pm"       "5.8"          "4,500"        "9.08am"       "4k"           "1.30pm"       "1.6m"        
 [526] "23rd"         "1962"         "1966"         "17th"         "14th"         "18-month"     "o2"          
 [533] "3.5m"         "2.23pm"       "21st-century" "19,000"       "5.2"          "20-year"      "2012-13"     
 [540] "12.00pm"      "7,500"        "2.50pm"       "ar-15"        "12.5"         "48-hour"      "1949"        
 [547] "4.8"          "3.25pm"       "9.21am"       "14bn"         "6.8"          "1.6bn"        "12.02pm"     
 [554] "6.4"          "6.2"          "26,000"       "240"          "140,000"      "26m"          "9.34am"      
 [561] "4.51pm"       "115"          "7.30pm"       "1.7m"         "2.9bn"        "3.17pm"       "9.15am"      
 [568] "2,400"        "450,000"      "9.40am"       "7.30"         "2gb"          "11bn"         "3.10pm"      
 [575] "6.9"          "7.20"         "10.12am"      "550"          "9.47am"       "16th"         "155"         
 [582] "10.37am"      "1am"          "3.15pm"       "13bn"         "2.30am"       "10.20am"      "9.45am"      
 [589] "2,700"        "9.19am"       "2.52pm"       "16bn"         "23,000"       "31,000"       "41-year-old" 
 [596] "1950"         "12-month"     "1939"         "9.16am"       "900m"         "4.7"          "12.31pm"     
 [603] "32-year-old"  "160m"         "55m"          "24,000"       "1928"         "1920s"        "6s"          
 [610] "350,000"      "220"          "2.47pm"       "18m"          "7.30am"       "9.41am"       "50-50"       
 [617] "12.36pm"      "3.1"          "2.40pm"       "2.56pm"       "28,000"       "1940"         "22bn"        
 [624] "9.36am"       "3.19pm"       "7.53am"       "4.52pm"       "4.29pm"       "5c"           "00"          
 [631] "10.11am"      "900,000"      "15-year"      "3.28pm"       "9.52am"       "8.50am"       "17m"         
 [638] "2.01pm"       "6.33pm"       "10.55am"      "4.2"          "7.52am"       "2.0"          "2008-09"     
 [645] "3.8bn"        "10-minute"    "2.45pm"       "6.3"          "750,000"      "43-year-old"  "12.11pm"     
 [652] "7.6"          "10.31am"      "9.39am"       "5.7"          "3.47pm"       "10.29am"      "4.15pm"      
 [659] "80bn"         "24-year-olds" "-1"           "25bn"         "25th"         "8.37am"       "1tn"         
 [666] "3.11pm"       "11.09am"      "9.06am"       "3.40pm"       "2018-19"      "1957"         "650"         
 [673] "8.00am"       "7.57am"       "11.33am"      "5,500"        "37,000"       "0.75"         "7-eleven"    
 [680] "5.42pm"       "10.5"         "7.50am"       "2.24pm"       "10.4"         "1,800"        "11.42am"     
 [687] "10.03am"      "23m"          "36-year-old"  "8.56am"       "5.4"          "8.08am"       "9.26am"      
 [694] "8.39am"       "5.06pm"       "4.4"          "9.29am"       "9.48am"       "11.10am"      "3.22pm"      
 [701] "19th-century" "12pm"         "105"          "8.35am"       "8.07am"       "7.51am"       "9.22am"      
 [708] "2.26pm"       "5.9"          "1.50"         "8.25am"       "4.2bn"        "45th"         "13th"        
 [715] "11.00am"      "13m"          "11.31am"      "4.13pm"       "85,000"       "2.4bn"        "12.09pm"     
 [722] "4.41pm"       "3.18pm"       "8.52am"       "172"          "9.17am"       "0800"         "125,000"     
 [729] "70-year-old"  "2.05pm"       "6.5bn"        "3.56pm"       "3.38pm"       "10.13am"      "65,000"      
 [736] "2.32pm"       "291"          "650,000"      "1800"         "10.25am"      "4.1"          "45bn"        
 [743] "109"          "8.22am"       "2.28pm"       "4.05pm"       "1944"         "143"          "10.08am"     
 [750] "3.08pm"       "10.47am"      "9.28am"       "1958"         "1941"         "1.45pm"       "5.3"         
 [757] "8.43am"       "4.03pm"       "1.36pm"       "8.05am"       "5p"           "5.05pm"       "9.30pm"      
 [764] "8.10am"       "10.24am"      "9.00am"       "9.4"          "2,800"        "1.49pm"       "360-degree"  
 [771] "7.59am"       "850,000"      "50th"         "5.37pm"       "33-year-old"  "12-year"      "1947"        
 [778] "8.09am"       "1.8m"         "12.32pm"      "8.42am"       "3.57pm"       "5.27pm"       "2.04pm"      
 [785] "1.54pm"       "8.12am"       "170,000"      "47-year-old"  "2.37pm"       "8.15am"       "10.36am"     
 [792] "1952"         "12.30am"      "580m"         "6.1"          "5.6"          "13.8"         "10.21am"     
 [799] "10.17am"      "q3"           "8.14am"       "10p"          "37-year-old"  "225"          "3.45pm"      
 [806] "8.49am"       "70bn"         "1954"         "2007-08"      "1946"         "1.25"         "3.51pm"      
 [813] "4.48pm"       "1.43pm"       "10.05am"      "450m"         "10.2"         "2.2bn"        "10.44am"     
 [820] "54,000"       "10.39am"      "12.22pm"      "8.19am"       "10.41am"      "9.42am"       "12.14pm"     
 [827] "10.34am"      "3.55pm"       "7.16pm"       "3.21pm"       "737"          "42-year-old"  "9.3"         
 [834] "2.14pm"       "8.24am"       "8.21am"       "4.02pm"       "20-point"     "1953"         "190"         
 [841] "227"          "8.45am"       "12.18pm"      "9.27am"       "10.06am"      "3.34pm"       "1.46pm"      
 [848] "2023"         "2.2m"         "1.1m"         "11-year"      "9.55am"       "3.27pm"       "3.33pm"      
 [855] "5.36pm"       "52.7"         "457"          "7.22pm"       "4.5m"         "1300"         "50-year-old" 
 [862] "10.43am"      "10.23am"      "147"          "11.44am"      "21m"          "2.21pm"       "4.6"         
 [869] "7.14pm"       "12.06pm"      "10.10am"      "4.34pm"       "4.09pm"       "1.48pm"       "8.17am"      
 [876] "320"          "1.27pm"       "5.22pm"       "200bn"        "220,000"      "1.0"          "36bn"        
 [883] "4.43pm"       "9.44am"       "8.04am"       "2.53pm"       "2.17pm"       "10.32am"      "enltr2"      
 [890] "5ft"          "102"          "7.48am"       "60bn"         "750m"         "10.45am"      "190,000"     
 [897] "6,500"        "5.01pm"       "mid-1990s"    "54-year-old"  "11.40am"      "8.13am"       "32.5"        
 [904] "6.00pm"       "11.13am"      "11.06am"      "10.02am"      "9.25am"       "5.56pm"       "8.30pm"      
 [911] "40th"         "3.31pm"       "8.48am"       "10-point"     "12.51pm"      "2.1bn"        "8.54am"      
 [918] "164"          "5.45pm"       "53.6"         "103"          "39-year-old"  "12.48am"      "6.28pm"      
 [925] "2,200"        "3,200"        "10.09am"      "8.55am"       "21bn"         "32,000"       "6.6"         
 [932] "9.57am"       "11.22am"      "9.33am"       "24m"          "8.06pm"       "5.11pm"       "20-minute"   
 [939] "7.15pm"       "g8"           "10.19am"      "11.45am"      "1.59pm"       "10.53pm"      "38-year-old" 
 [946] "9.5"          "2035"         "2,300"        "9.50am"       "10.30pm"      "1.44pm"       "4.44pm"      
 [953] "4.25pm"       "1.31pm"       "12.27pm"      "6.24pm"       "3.53pm"       "1.07am"       "1.30am"      
 [960] "145"          "118"          "10.04am"      "1932"         "8.38am"       "45m"          "12.07pm"     
 [967] "10.56am"      "8.5"          "4.11pm"       "9.43am"       "2.03pm"       "6.36pm"       "4.45pm"      
 [974] "18-"          "11.19am"      "10.38am"      "10.07am"      "6.20pm"       "6.12pm"       "12.28pm"     
 [981] "12.25pm"      "2,600"        "11.27am"      "12.47pm"      "9.18am"       "72,000"       "160bn"       
 [988] "under-18s"    "10.45pm"      "4.20pm"       "9.05am"       "7.20pm"       "12.45pm"      "8.47am"      
 [995] "1.75"         "75m"          "11.5"         "4.16pm"       "11.34am"      "2.1m" 
```

**Comment by kbenoit** [comment]
"word" separates currency symbols true, which is why I added this test:
```r
txt <- "1 and 12 123 1975 12345 100,000 $1,000.00 123,123,456 and 50¢ 1.200,34"
toks2 <- tokens(txt, what = "fasterword", remove_numbers = TRUE)
```
since the second will not separate them. My view is that we would want to remove everything above as a "number" EXCEPT:
- anything plural, e.g. "1990s", "30s" - these are arguably not "numbers" and are also just make a complicated regex more complicated
- times, e.g. "3:50pm" or "15:50"
- phrases starting with numerals, such as "34-year-old", "19th-century"
- distances or units combined with numbers, such as 100Mb, 50m

but we probably DO want to remove
- 100bn
- ordinal numbers, e.g. "1st" - but happy not to since these are adding complexity

We could just keep it simple and add some canned regexes for removal, for specialist removals, such as ordinal numbers, etc.

**Comment by koheiw** [comment]
It is probably impossible to find a pattern that makes everyone happy, so how about just removing pure numeric tokens that match `^[,.\\p{N}]+$`? We don't need special handing for currencies then.

**Comment by kbenoit** [comment]
Fine with me.

**Comment by kbenoit** [comment]
I went with this one:

``` r
txt <- unlist(stringi::stri_split_fixed("1 and 12 123 1975 12345 100,000 $1,000.00 123,123,456 and 50¢ 1.200,34 100bn 20-year-old", " "))
txt
##  [1] "1"           "and"         "12"          "123"         "1975"       
##  [6] "12345"       "100,000"     "$1,000.00"   "123,123,456" "and"        
## [11] "50¢"         "1.200,34"    "100bn"       "20-year-old"
stringi::stri_subset_regex(txt, "^\\p{Sc}{0,1}\\p{N}+([.,]*\\p{N})*(bn|mn|\\p{Sc}){0,1}$", negate = TRUE)
## [1] "and"         "and"         "20-year-old"
```

**Comment by koheiw** [comment]
I am fine with the Unicode currency sign, but let's not include language specific patter `bn|mn`.

**Pull request merged**